File: elixir/day1.exs
defmodule Day1 do
  # Find first and last integer and return the 2-digit number formed
  def find_cal_value line do
    #  Replace all letters with ""
    digits = Regex.replace(~r/[^0-9]/, line, "") |> String.graphemes()
    case digits do
      #  Default case
      [] ->
        0
      # Get first and last, combine and return
      _ ->
        first = digits |> List.first() |> String.to_integer()
        last = digits |> List.last() |> String.to_integer()
        first * 10 + last
    end
  end
end

part1 = fn ->
  # Read input, split on line and map to values -> reduce to sum
  with {:ok, body} <- File.read("../inputs/day1.txt") do
    String.split(body, "\n")
    |> Enum.map(& Day1.find_cal_value(&1))
    |> Enum.reduce(& &1 + &2)
  else
    {:error, reason} ->
      IO.puts reason
      exit :shutdown
  end
end

part2 = fn ->
  replace_numstrings = fn value ->
    regex = Regex.compile!("(one|two|three|four|five|six|seven|eight|nine)")
    mapping = %{"one" => "1", "two" => "2", "three" => "3", "four" => "4", "five" => "5", "six" => "6", "seven" => "7", "eight" => "8", "nine" => "9"}
    value
    |> String.graphemes()
    |> Enum.reduce("", fn char, acc ->
      chunk = acc <> char
      case Regex.run(regex, chunk) do
        nil ->
          chunk
        match ->
          num = hd(match)
          pre = num |> String.graphemes() |> hd()
          post = num |> String.graphemes() |> Enum.reverse() |> hd()
          Regex.replace(regex, chunk, pre <> Map.get(mapping, num) <> post)
      end
    end)
  end

  # Read input, split by line, replace text numbers -> reduce to sum
  with {:ok, body} <- File.read("../inputs/day1.txt") do
    String.split(body, "\n")
    |> Enum.map(& replace_numstrings.(&1))
    |> Enum.map(& Day1.find_cal_value(&1))
    |> Enum.reduce(& &1 + &2)
  else
    {:error, reason} ->
      IO.puts reason
      exit :shutdown
    end
end



IO.puts part1.()

IO.puts part2.()

File: python/day1.py
def convert_cal_value(calibration_value: str) -> str:
    mapping = {"one":"1", "two":"2", "three":"3", "four":"4", "five":"5", "six":"6", "seven":"7", "eight":"8", "nine":"9"}
    start = 0
    i = 1
    while [calibration_value for key in mapping.keys() if(key in calibration_value)]:
        test = calibration_value[start:i]
        if [test for key in mapping.keys() if(key in test)]:
            test = [key for key in mapping.keys() if(key in test)][0]
            calibration_value = calibration_value.replace(test, mapping[test])
            start = i - len(test) + 1
            i = start
        i += 1

    return calibration_value

def find_cal_val(calibration_value):
    first = last = "0"
    length = len(calibration_value)
    for i in range(len(calibration_value)):
        if first == "0" and calibration_value[i] in "123456789":
            first = calibration_value[i]
        if last == "0" and calibration_value[length - i - 1] in "123456789":
            last = calibration_value[length - i - 1]
    return int(first + last)

def main():
    values = open("../inputs/day1.txt").read().split("\n")
    total = 0
    for value in values:
        total += find_cal_val(value)
    print("Part 1:", total)

    total = 0
    for value in values:
        total += find_cal_val(convert_cal_value(value))
    print("Part 2:", total)

if __name__ == "__main__":
    main()